File: 224_basic_calculator.py
class Solution:
    def calculate(self, s):
        """
            https://leetcode.com/problems/basic-calculator/
            Time Complexity - O(n)
            Space Complexity - O(n)
            'n' is the length of 's'
        """
        result, num, sign = 0, 0, 1
        stack = []
        for c in s:
            if c.isdigit():
                num = 10 * num + int(c)
            elif c == '+':
                result += sign * num
                num = 0
                sign = 1
            elif c == '-':
                result += sign * num
                num = 0
                sign = -1
            elif c == "(":
                stack.append(result)
                stack.append(sign)
                sign = 1
                result = 0
            elif c == ")":
                result += sign * num
                result *= stack.pop()
                result += stack.pop()
                num = 0
        return result + num * sign


if __name__ == '__main__':
    print(Solution().calculate(" 2-1 + 2 "))

File: 224. Basic Calculator.java
class Solution {// Time and space of O(N)
    int result= 0;
    int idx = 0;
    int num = 0;
    int min = Integer.MIN_VALUE;
    int sign = 1 ; // 1 for positive , -1 for negative
    public int calculate(String s) {
      //Base case 
        if(s.length() == 0)
            return 0;
        Stack<Integer> stack = new Stack<>(); 
        
        while(idx < s.length()){
            char c  = s.charAt(idx);
           // System.out.println(c);
            {
             if(Character.isDigit(c)){
                    num = num*10 + (c -'0');
                    
                }
             if(!Character.isDigit(c) && c != ' '|| idx == s.length() - 1){
                 
                 if(c == '('){
                     stack.push(sign);
                     //System.out.print(sign);
                     sign = 1;
                     stack.push(min);
                     
                 }
                else if(sign == 1){
                     stack.push(num);
                     }
                 else if(sign == -1){
                     stack.push(-num);
                     }
                 if(c == ')' ){
                 while(!stack.isEmpty() && stack.peek() != min){
                     result += stack.pop();
                 }
                 stack.pop();
                 stack.push(stack.pop()*result);                 
             }
                num = 0 ;
                result = 0;
                
              //" 2-1 + 2 "  
                 if(c== '+'|| c =='-'){
                     sign = c=='+'? 1:-1 ;
                 }
            }
            idx++;
          }
        }
        
        while (!stack.isEmpty() && stack.peek() != min ){
            result += stack.pop();
        }
        return result;
    }
   
}
//"   (  3 ) "   "(1+(4+5+2)-3)+(6+8)   "2-(5-6)""